### PR TITLE
Improve adaptive solvers for density computation

### LINKED ISSUE #739: Failing tests for diffusion networks on Mac
**Describe the bug**

Some tests are failing on Mac. 

**To Reproduce**
Minimal steps to reproduce the behavior:
1. On Mac Silicon
2. Checkout and setup bayesflow (tensorflow or torch as a backend)
3. run `pytest tests/test_networks/test_inference_networks.py::test_density_numerically`
4. The following tests fail:

```
tests/test_networks/test_inference_networks.py::test_density_numerically[batch_size=2-feature_size=2-conditions_size=None-generative_inference_network='diffusion_model_cosine_velocity'] FAILED 
tests/test_networks/test_inference_networks.py::test_density_numerically[batch_size=2-feature_size=2-conditions_size=None-generative_inference_network='diffusion_model_edm_potential'] FAILED  
```
**Expected behavior**

Tests should pass. To me, it is unclear whether there is an issue with the code or with the test.

Just a thought: Perhaps it wouldn't hurt to run automatic tests on GitHub with Mac. We would probably detect this issue sooner.

**Minimality**
- [x] I verify that my example is minimal, does not rely on third-party packages, and is most likely an issue in BayesFlow.

### PR DESCRIPTION
**Pull request by arrjon**
Trying to fix #739 with improved adaptive solvers for density computation. 

I tightened the default tolerances for density computations (in general multi-component states). Moreover, I centralized error normalization for computing adaptive step sizes, which is important if multiple states (`xz` and `trace` for density computation) are passed.

### CODE REVIEW

**Comment by paul-buerkner** [comment]
Thank you for working on this! I tried with, I think, your latest version of the PR and tests are still failing for me. Given your commits, I take it things are failing for you too still?

**Comment by arrjon** [comment]
I am trying to find the best overall configuration of the adaptive sampler. At least with the current version, tests pass locally for me and in the CI. Only with the docs is something wrong, which is unrelated to my changes, I think.

**Comment by arrjon** [comment]
I tried to fix the docs building; it seems to be something with `sphinx-polyversion 3.0.0` dropping `buffer_size`.

**Comment by stefanradev93** [comment]
Something is wrong with the new tests. They are too slow under tensorflow, which probably reveals a suboptimal implementation from our side.

**Comment by arrjon** [comment]
Indeed, the tighter tolerances can lead to many more steps in the ODE solver, so a bit longer runtime is expected but not a timeout...

**Comment by stefanradev93** [comment]
Yes, this is probably a pointer that the corresponding backend-agnostic autodiff utility under the tensorflow backend is suboptimal.

**Comment by arrjon** [comment]
I think `tensorflow` was not compiling before, which was the reason for the long runs. Now it should. I tuned the tolerances such that they also work on a GPU now.

**Comment by arrjon** [comment]
I reorganized the tests around diffusion models to reduce test time while trying to keep the coverage. Please have a look at it.

**Comment by paul-buerkner** [comment]
Somehow the tests indeed run much much faster now on my mac. And then do pass too, which is nice :-D

Thank you for working on it! @stefanradev93 what do you think about these fixes?

**Comment by arrjon** [comment]
The main bottleneck was computing the numerical Jacobian using an adaptive solver on untrained networks (with high error and many adaptive steps). Density computation with adaptive solvers is now checked in a designated test.

**Comment by stefanradev93** [comment]
These fixes are great, thanks for working on it @arrjon! Some other things regarding compositional sampling will become easier once #701 is merged with keras-based adapetrs.
